EventedPLEG: Pass event created timestamp correctly to cache

hshiina opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind bug

#### What this PR does / why we need it:
`CreatedAt` timestamp of `ContainerEventResponse` should be passed as nanoseconds to `time.Unix()`.

This fix may mitigate #123087. I guess this issue happens as following:
Usually, a container created event is delivered while a pod worker is creating a pod at [SyncPod()](https://github.com/kubernetes/kubernetes/blob/v1.30.0-rc.2/pkg/kubelet/pod_workers.go#L1283) and the `created` status is cached by the PLEG. After `SyncPod()`, the pod worker waits at [GetNewerThan()](https://github.com/kubernetes/kubernetes/blob/v1.30.0-rc.2/pkg/kubelet/pod_workers.go#L1253) until a next event (container started) happens because the cached container status is older than `lastSyncTime`. However, due to the bug in this PR, the `modified` time of the cached status is so much newer because a nanoseconds value is set as seconds. Then, the pod worker gets a `created` container status from the cache and goes into `SyncPod()` soon. Then, the worker creates another container.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

hshiina committed:
Pass event created timestamp correctly to cache

`CreatedAt` timestamp of `ContainerEventResponse` should be passed as
nanoseconds to `time.Unix()`.

bart0sh commented:
/triage accepted
/priority important-longterm
/ok-to-test
/cc

bart0sh commented on pkg/kubelet/pleg/evented.go:
```diff
@@ -264,7 +264,7 @@ func (e *EventedPLEG) processCRIEvents(containerEventsResponseCh chan *runtimeap
 			shouldSendPLEGEvent = true
 		} else {
-			if e.cache.Set(podID, status, err, time.Unix(event.GetCreatedAt(), 0)) {
+			if e.cache.Set(podID, status, err, time.Unix(0, event.GetCreatedAt())) {
```
@hshiina great catch! Thank you!
/lgtm

SergeyKanzhelev replied:
nice:

- CRI-O: https://github.com/cri-o/cri-o/blob/36c2750916eb9e75f01a627463c2cfad207f192d/server/server.go#L934
- Containerd: https://github.com/containerd/containerd/blob/f5d5407c2ff12865653a9a132d5783196be82763/internal/cri/server/helpers.go#L393

Both return nano indeed

bart0sh commented:
/assign @mrunalp @derekwaynecarr @dchen1107 
for approval

rphillips commented:
/test pull-kubernetes-e2e-kind-evented-pleg
/test pull-kubernetes-e2e-kind-beta-features
/test pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e

rphillips commented:
cc @harche

rphillips commented:
failure looks unrelated... retesting

/test pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e

sairameshv commented:
The [failure](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/124297/pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e/1787930420118032384) seems to be expected as described in [this](https://github.com/kubernetes/kubernetes/issues/124704) bug
```
E2eNode Suite: [It] [sig-node] [NodeFeature:SidecarContainers] Containers Lifecycle should not hang in termination if terminated during initialization expand_less	14s
{ failed [FAILED] should delete in < 5 seconds, took 8.044552
Expected
    <float64>: 8.044551982
to be <
    <int64>: 5
In [It] at: k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:2961 @ 05/07/24 20:11:30.288
}
```

hshiina committed:
Update global cache timestamp more frequently

There are some cases where a pod worker is woken up without a cache
update by the PLEG such as a pod termination. Then, the worker gets
stuck in `cache.GetNewerThan()` till the global cache timestamp is
updated by the PLEG. In order to unblock the stuck worker as early as
the Generic PLEG, this fix makes the Evented PLEG update the global
cache as frequently as the Generic PLEG.

harche commented:
/cc @pacoxu

hshiina commented:
A failure in [pull-kubernetes-node-e2e-containerd](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/124297/pull-kubernetes-node-e2e-containerd/1788130137837932544) seems to be caused because a regular container and a postStart hook start asynchronously.

```
{ failed [FAILED] couldn't find that PostStart-regular-1 ever started, got
0) 2024-05-08 09:07:01.942 +0000 UTC init-1 Starting
1) 2024-05-08 09:07:01.979 +0000 UTC init-1 Started
2) 2024-05-08 09:07:02.007 +0000 UTC init-1 Delaying
3) 2024-05-08 09:07:03.023 +0000 UTC init-1 Exiting
4) 2024-05-08 09:07:03.729 +0000 UTC PostStart-regular-1 Starting
5) 2024-05-08 09:07:03.735 +0000 UTC regular-1 Starting
6) 2024-05-08 09:07:03.754 +0000 UTC PostStart-regular-1 Started
7) 2024-05-08 09:07:03.755 +0000 UTC regular-1 Started
8) 2024-05-08 09:07:03.77 +0000 UTC regular-1 Delaying
9) 2024-05-08 09:07:03.771 +0000 UTC PostStart-regular-1 Delaying
10) 2024-05-08 09:07:04.789 +0000 UTC PostStart-regular-1 Exiting
11) 2024-05-08 09:07:13.862 +0000 UTC regular-1 Exiting
In [It] at: k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:286 @ 05/08/24 09:07:17.158
}
```

Because of #124668, `RunTogether()` requires a regular container start before a postStart hook. It might be better to add `StartDelay` here:
https://github.com/kubernetes/kubernetes/blob/8e5b26b4c373fcbb9b445a2157a505aeaf7ab69f/test/e2e_node/container_lifecycle_test.go#L255-L259

hshiina commented:
A failure in [pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/124297/pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e/1788130523395133440) looks the same issue described in #124743.

```
Expected
    <int64>: 29479
to be within 1000 of ~
    <int>: 32000
In [It] at: k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:2876 @ 05/08/24 09:19:49.794
```

hshiina commented:
/test pull-kubernetes-node-e2e-containerd

hshiina commented:
/test pull-kubernetes-e2e-kind-beta-features

hshiina commented:
/test pull-kubernetes-e2e-kind-evented-pleg

hshiina commented:
/test pull-kubernetes-e2e-kind-evented-pleg

hshiina commented:
One of failures in [pull-kubernetes-e2e-kind-evented-pleg](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/124297/pull-kubernetes-e2e-kind-evented-pleg/1792131661371019264) looks the same issue as #123087:

```
{ failed [FAILED] 2 containers failed which is more than allowed 1
In [It] at: k8s.io/kubernetes/test/e2e/network/service.go:1270 @ 05/19/24 10:14:07.319
}
```


I guess the bind error in the [log](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/124297/pull-kubernetes-e2e-kind-evented-pleg/1792131661371019264/build-log.txt) was caused because two containers were created.

```
  I0519 10:14:07.285073   66190 runners.go:195] Running kubectl logs on non-ready containers in services-1503
  I0519 10:14:07.318684 66190 kubectl_utils.go:128] Logs of services-1503/nodeport-test-xwqf5:nodeport-test on node kind-worker2
  I0519 10:14:07.318727 66190 kubectl_utils.go:129]  : STARTLOG
  I0519 10:14:03.496089   	1 log.go:245] Error from net.Listen(): listen tcp :9376: bind: address already in use

  ENDLOG for container services-1503:nodeport-test-xwqf5:nodeport-test
  I0519 10:14:07.318794 66190 service.go:1270] Unexpected error:
  	<*errors.errorString | 0xc00128ecf0>:
  	2 containers failed which is more than allowed 1
  	{
      	s: "2 containers failed which is more than allowed 1",
  	}
  [FAILED] in [It] - k8s.io/kubernetes/test/e2e/network/service.go:1270 @ 05/19/24 10:14:07.319
```

I don’t think it is enough to wait for the cache to be updated once at `GetNewerThan()` without confirming if it is updated as expected. After a pod worker starts a container in `SyncPod()`, if the pod worker enters syncPod() again before a container started event is delivered, the worker tries to create the container again. Because pod workers do not remember what they requested the container runtime to do, they compute actions in `SyncPod()` only by seeing cached container statuses. So, before entering `SyncPod()` again, cached statuses of all containers that were requested to start need to be updated by PLEG. However, just calling `cache.GetNewerThan()` once does not guarantee that all events are received by PLEG, especially in a case of Evented PLEG where events are delivered independently.

hshiina commented:
I posted another PR (#124953) to try solving the remaining problem.

I think it's still worth merging this PR at first in order to improve the Evented PLEG even though there is still an unsolved problem.

pacoxu commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features

hshiina commented:
/retest

SergeyKanzhelev commented on pkg/kubelet/pleg/evented.go:
```diff
@@ -34,7 +34,7 @@ import (
 // is to be updated periodically. If pod workers get stuck at cache.GetNewerThan
 // call, after this period it will be unblocked.
-const globalCacheUpdatePeriod = 5 * time.Second
+const globalCacheUpdatePeriod = 1 * time.Second
```
is this change part of this PR?

SergeyKanzhelev replied:
I would expect the comment change here to highlight why it is 1, so next time somebody decide to change it will have some reference on why it is 1 and not something else

hshiina replied:
> is this change part of this PR?

With fixing the timestamp issue, we hit a failure in CI as described in https://github.com/kubernetes/kubernetes/pull/124297#issuecomment-2099637107 (It took too much time to delete a pod). The detail of this issue is described in #124704. In order to pass tests with Evented PLEG enabled, I added this fix for `globalCacheUpdatePeriod` into this PR.

hshiina replied:
> I would expect the comment change here to highlight why it is 1, so next time somebody decide to change it will have some reference on why it is 1 and not something else

Thank you for your feedback. I modified this PR with explaining that the cache update period should not be longer than the relisting period of Generic PLEG.

SergeyKanzhelev replied:
I am super confused why this PR mixes change of relist timeout with the fixing the time handling.

hshiina committed:
Address feedback

hshiina commented:
/test pull-kubernetes-e2e-kind-evented-pleg
/test pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e

hshiina commented:
/retest

HirazawaUi commented on pkg/kubelet/pleg/evented.go:
```diff
@@ -125,7 +125,7 @@ func (e *EventedPLEG) Start() {
 	e.stopCacheUpdateCh = make(chan struct{})
 	go wait.Until(e.watchEventsChannel, 0, e.stopCh)
-	go wait.Until(e.updateGlobalCache, globalCacheUpdatePeriod, e.stopCacheUpdateCh)
+	go wait.Until(e.updateGlobalCache, e.globalCacheUpdatePeriod, e.stopCacheUpdateCh)
```
After we modify the timestamp from seconds to nanoseconds, do we really need to execute `updateGlobalCache()` at a faster interval?

`updateGlobalCache()` serves as a last resort for recovery from failures. We shouldn’t perform this action for any pod that can retrieve podStatus normally (because it is not the actual timestamp returned by the runtime), 

The 5-second interval comes from the timeout set in [NewRemoteRuntimeService](https://github.com/kubernetes/kubernetes/blob/f5c538418189e119a8dbb60e2a2b22394548e326/staging/src/k8s.io/cri-client/pkg/remote_runtime.go#L114-L117). Unless we modify the timeout of `NewRemoteRuntimeService` to 1 second and assume that if the runtime fails to respond within 1 second, it should be considered in a failure state.

Then, we can use `updateGlobalCache()` for failure recovery.

hshiina replied:
Pod workers are expected to be unblocked from `GetNewerThan()` in one second at pod termination:
https://github.com/kubernetes/kubernetes/blob/f5c538418189e119a8dbb60e2a2b22394548e326/pkg/kubelet/pod_workers.go#L1247-L1256

Generic PLEG also updates the global cache every second:
https://github.com/kubernetes/kubernetes/blob/f5c538418189e119a8dbb60e2a2b22394548e326/pkg/kubelet/pleg/generic.go#L356

However, it might better to remove `GetNewerThan()` from `podWorkerLoop()` as mentioned at https://github.com/kubernetes/kubernetes/issues/124704#issuecomment-2336712682. I wonder if I would revert the latter commit in order to fix the apparent bug at first.

HirazawaUi replied:
> Generic PLEG also updates the global cache every second:

This might be because, at this point, we have ensured that all pods in the cache have been correctly updated.

@Yujuhong, is my understanding correct? If not, please correct me.

However, for the current updateGlobalCache(), we don’t have such a guarantee, so the same strategy shouldn’t be applied.



> However, it might better to remove `GetNewerThan()` from `podWorkerLoop()` as mentioned at [#124704 (comment)](https://github.com/kubernetes/kubernetes/issues/124704#issuecomment-2336712682). I wonder if I would revert the latter commit in order to fix the apparent bug at first.

You might want to only fix the timestamp. A more consistent approach would be to also fix: https://github.com/kubernetes/kubernetes/blob/f5c538418189e119a8dbb60e2a2b22394548e326/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L1583-L1589

hshiina replied:
> You might want to only fix the timestamp. A more consistent approach would be to also fix:
> 
> https://github.com/kubernetes/kubernetes/blob/f5c538418189e119a8dbb60e2a2b22394548e326/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L1583-L1589

This looks like a different issue (#126414) to me because the runtime (CRI-O) sets this timestamp in seconds. This should be fixed in both kubernetes and CRI-O synchronously.

HirazawaUi replied:
Making this change does not affect our existing behavior. Referring to the following code, given a unix timestamp in seconds, whether it is kept in seconds or converted to nanoseconds, the behavior will remain consistent.
```
func main() {
	if !compare1() && !compare2() {
		fmt.Printf("Their behavior is consistent")
	}
}

func compare2() bool {
	eventedTime := time.Now().UnixNano()
	time.Sleep(100 * time.Millisecond)
	// The initial timestamp is seconds
	genericTime := time.Now().Unix()

	cachedEventedTime := time.Unix(0, eventedTime)
	// Convert to nanoseconds
	cachedGenericTime := time.Unix(0, genericTime)

	if cachedEventedTime.Before(cachedGenericTime) {
		fmt.Printf("Expected: evented:%v, generic:%v\n", cachedEventedTime, cachedGenericTime)
		return true
	} else {
		fmt.Printf("Unexpected: evented:%v, generic:%v\n", cachedEventedTime, cachedGenericTime)
		return false
	}
}

func compare1() bool {
	eventedTime := time.Now().UnixNano()
	time.Sleep(100 * time.Millisecond)
	// The initial timestamp is seconds
	genericTime := time.Now().Unix()

	cachedEventedTime := time.Unix(0, eventedTime)
	// Keep in seconds
	cachedGenericTime := time.Unix(genericTime, 0)

	if cachedEventedTime.Before(cachedGenericTime) {
		fmt.Printf("Expected: evented:%v, generic:%v\n", cachedEventedTime, cachedGenericTime)
		return true
	} else {
		fmt.Printf("Unexpected: evented:%v, generic:%v\n", cachedEventedTime, cachedGenericTime)
		return false
	}
}

```
Therefore, we can first fix kubelet and then fix CRI-O to ensure that this functionality works properly in the end.

SergeyKanzhelev replied:
it looks to me that CRIO returns nano: https://github.com/cri-o/cri-o/blob/36c2750916eb9e75f01a627463c2cfad207f192d/server/server.go#L934 ? Am I missing something?

HirazawaUi replied:
The `PodSandboxStatus` interface in CRI-O returns the timestamp in seconds, and kubelet uses this timestamp in `genericPLEG`.

crio:
https://github.com/cri-o/cri-o/blob/11934304d11256189df8e316e195af6650a472aa/server/sandbox_status.go#L43-L48

kubelet:
https://github.com/kubernetes/kubernetes/blob/342da505bdefbd849b808cca3cb76c24a993025f/pkg/kubelet/pleg/generic.go#L445-L482

I have already filed an issue in the CRI-O project, and it should be fixed soon, ref: https://github.com/cri-o/cri-o/issues/8580

hshiina committed:
Revert "Address feedback"

This reverts commit 7aa7062892a8084725b53a6f3395b76245cb94a7.

hshiina committed:
Revert "Update global cache timestamp more frequently"

This reverts commit dbc47341fabfc6dcd6b6bc0854f0a59ce4095827.

hshiina commented:
Please let me start over.

/test pull-kubernetes-e2e-kind-evented-pleg
/test pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e

hshiina commented:
Now, only Evented PLEG caching is being fixed.

The jobs with EventedPLEG enabled were almost passed because this fix mitigates #123087. (Conversely, this would make it difficult to reproduce #123087.)

In addition, another issue appeared with by fix. There was a failure in the [job](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/124297/pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e/1833419918532415488):
```
E2eNode Suite: [It] [sig-node] [NodeFeature:SidecarContainers] Containers Lifecycle [It] should not hang in termination if terminated during initialization [sig-node, NodeFeature:SidecarContainers]
{ failed [FAILED] should delete in < 5 seconds, took 10.071795
Expected
    <float64>: 10.071794756
to be <
    <int64>: 5
In [It] at: k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:3470 @ 09/10/24 08:54:31.819
}
```

With this fix, a pod worker is blocked at `GetNewerThan()`:
https://github.com/kubernetes/kubernetes/blob/d913914511c87dee288d1d72921c364a8331893f/pkg/kubelet/pod_workers.go#L1256

We can see a pod worker blocked at pod termination in [kubelet.log](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/124297/pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e/1833419918532415488/artifacts/tmp-node-e2e-d4b0ca81-fedora-coreos-40-20240808-3-0-gcp-x86-64/kubelet.log). This caused the failure:

```
I0910 08:54:21.758046	4615 kubelet.go:2423] "SyncLoop DELETE" source="api" pods=["containers-lifecycle-test-6189/dont-hang-if-terminated-in-init"]
I0910 08:54:21.758096	4615 pod_workers.go:854] "Pod is marked for graceful deletion, begin teardown" pod="containers-lifecycle-test-6189/dont-hang-if-terminated-in-init" podUID="41a59818-90ea-4b92-8b33-92b9afda42f7" updateType="update"
I0910 08:54:21.758123	4615 pod_workers.go:963] "Notifying pod of pending update" pod="containers-lifecycle-test-6189/dont-hang-if-terminated-in-init" podUID="41a59818-90ea-4b92-8b33-92b9afda42f7" workType="terminating"
I0910 08:54:21.758145	4615 pod_workers.go:970] "Cancelling current pod sync" pod="containers-lifecycle-test-6189/dont-hang-if-terminated-in-init" podUID="41a59818-90ea-4b92-8b33-92b9afda42f7" workType="terminating"
I0910 08:54:21.758180	4615 pod_workers.go:1233] "Processing pod event" pod="containers-lifecycle-test-6189/dont-hang-if-terminated-in-init" podUID="41a59818-90ea-4b92-8b33-92b9afda42f7" updateType="terminating"

<The pod worker was being blocked at `GetNewerThan()`>

I0910 08:54:25.234136    4615 kubelet.go:2006] "SyncTerminatingPod enter" pod="containers-lifecycle-test-6189/dont-hang-if-terminated-in-init" podUID="41a59818-90ea-4b92-8b33-92b9afda42f7"

<snip: The pod worker ran `SyncTermninatingPod()`>

I0910 08:54:25.736947	4615 kubelet.go:2093] "SyncTerminatingPod exit" pod="containers-lifecycle-test-6189/dont-hang-if-terminated-in-init" podUID="41a59818-90ea-4b92-8b33-92b9afda42f7"
I0910 08:54:25.736961	4615 pod_workers.go:1402] "Pod terminated all containers successfully" podUID="41a59818-90ea-4b92-8b33-92b9afda42f7"
I0910 08:54:25.736962	4615 status_manager.go:227] "Syncing updated statuses"
I0910 08:54:25.737009	4615 pod_workers.go:1513] "Pending update already queued" podUID="41a59818-90ea-4b92-8b33-92b9afda42f7"
I0910 08:54:25.737031	4615 pod_workers.go:1338] "Processing pod event done" pod="containers-lifecycle-test-6189/dont-hang-if-terminated-in-init" podUID="41a59818-90ea-4b92-8b33-92b9afda42f7" updateType="terminating"
I0910 08:54:25.737049	4615 pod_workers.go:1233] "Processing pod event" pod="containers-lifecycle-test-6189/dont-hang-if-terminated-in-init" podUID="41a59818-90ea-4b92-8b33-92b9afda42f7" updateType="terminated"

<The pod worker was being blocked at `GetNewerThan()`>

I0910 08:54:30.235579	4615 kubelet.go:2151] "SyncTerminatedPod enter" pod="containers-lifecycle-test-6189/dont-hang-if-terminated-in-init" podUID="41a59818-90ea-4b92-8b33-92b9afda42f7"
```
This issue is addressed in #124704. I shouldn't have mixed it in this PR. It would be better to fix it separately.

HirazawaUi commented:
Due to the rapid action of the CRI-O team, this bug has already been fixed and backported to older versions, which means you can apply the fix.
https://github.com/kubernetes/kubernetes/blob/f5c538418189e119a8dbb60e2a2b22394548e326/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L1583-L1589

ref: https://github.com/cri-o/cri-o/pull/8582

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

thank you!

SergeyKanzhelev commented:
/retest
